test/Dialect/SPIRV/canonicalize.mlir

```diff
@@ -1,5 +1,111 @@
+// spv.AccsessChain
+//===----------------------------------------------------------------------===//
+
+func @access_chain_canonicalization_full_0() -> f32 {
+  // CHECK: %[[INDEX_0:.*]] = spv.constant 0 : i32
+  // CHECK-NEXT: %[[COMPOSITE:.*]] = spv.Variable : !spv.ptr<!spv.struct<!spv.array<4 x !spv.array<4 x f32>>, !spv.array<4 x i32>>, Function>
+  // CHECK-NEXT: %[[COMPONENT_0:.*]] = spv.AccessChain %[[COMPOSITE]][%[[INDEX_0]], %[[INDEX_0]], %[[INDEX_0]]] : !spv.ptr<!spv.struct<!spv.array<4 x !spv.array<4 x f32>>, !spv.array<4 x i32>>, Function>
+  // CHECK-NEXT: %[[RESULT:.*]] = spv.Load "Function" %[[COMPONENT_0]] : f32
+  // CHECK-NEXT: spv.ReturnValue %[[RESULT]] : f32
+  %c0 = spv.constant 0: i32
+  %0 = spv.Variable : !spv.ptr<!spv.struct<!spv.array<4x!spv.array<4xf32>>, !spv.array<4xi32>>, Function>
+  %1 = spv.AccessChain %0[%c0] : !spv.ptr<!spv.struct<!spv.array<4x!spv.array<4xf32>>, !spv.array<4xi32>>, Function>
+  %2 = spv.AccessChain %1[%c0] : !spv.ptr<!spv.array<4x!spv.array<4xf32>>, Function>
+  %3 = spv.AccessChain %2[%c0] : !spv.ptr<!spv.array<4xf32>, Function>
+  %5 = spv.Load "Function" %3 : f32
+  spv.ReturnValue %5 : f32
+}
+
+// -----
+
+func @access_chain_canonicalization_full_1() -> i32 {
+  // CHECK: %[[INDEX_0:.*]] = spv.constant 0 : i32
+  // CHECK-NEXT: %[[INDEX_1:.*]] = spv.constant 1 : i32
+  // CHECK-NEXT: %[[COMPOSITE:.*]] = spv.Variable : !spv.ptr<!spv.struct<!spv.array<4 x !spv.struct<!spv.array<4 x i32>>>>, Function>
+  // CHECK-NEXT: %[[COMPONENT:.*]] = spv.AccessChain %[[COMPOSITE]][%[[INDEX_0]], %[[INDEX_1]], %[[INDEX_0]], %[[INDEX_1]]] : !spv.ptr<!spv.struct<!spv.array<4 x !spv.struct<!spv.array<4 x i32>>>>, Function>
+  // CHECK-NEXT: %[[RESULT:.*]] = spv.Load "Function" %[[COMPONENT]] : i32
+  // CHECK-NEXT: spv.ReturnValue %[[RESULT]] : i32
+  %c0 = spv.constant 0: i32
+  %c1 = spv.constant 1: i32
+  %0 = spv.Variable : !spv.ptr<!spv.struct<!spv.array<4x!spv.struct<!spv.array<4xi32>>>>, Function>
+  %1 = spv.AccessChain %0[%c0, %c1] : !spv.ptr<!spv.struct<!spv.array<4x!spv.struct<!spv.array<4xi32>>>>, Function>
+  %2 = spv.AccessChain %1[%c0, %c1] : !spv.ptr<!spv.struct<!spv.array<4xi32>>, Function>
+  %3 = spv.Load "Function" %2 : i32
+  spv.ReturnValue %3 : i32
+}
+
+// -----
+
+// The result of the second `spv.AccessChain` operation has two uses and therefore cannot be combined with the third.
+func @access_chain_canonicalization_partial() -> !spv.array<4xf32> {
+  // CHECK: %[[INDEX_0:.*]] = spv.constant 0 : i32
+  // CHECK-NEXT: %[[COMPOSITE:.*]] = spv.Variable : !spv.ptr<!spv.struct<!spv.array<4 x !spv.array<4 x f32>>, !spv.array<4 x i32>>, Function>
+  // CHECK-NEXT: %[[COMPONENT_0:.*]] = spv.AccessChain %[[COMPOSITE]][%[[INDEX_0]], %[[INDEX_0]]] : !spv.ptr<!spv.struct<!spv.array<4 x !spv.array<4 x f32>>, !spv.array<4 x i32>>, Function>
+  // CHECK-NEXT: %[[COMPONENT_1:.*]] = spv.AccessChain %[[COMPONENT_0]][%[[INDEX_0]]] : !spv.ptr<!spv.array<4 x f32>, Function>
```

Comment:
We should also be able to turn this into having two indices if looking at the previous access chain op instead of the next one.